### ISSUE TITLE
Gen2.1.20 : ClassyWalk: Motor does not spin, can't detect SELF_HOLD, BUTTON, and hall order

### ISSUE THREAD

**Issue by hoverboardhavoc** [RoboDurden/Hoverboard-Firmware-Hack-Gen2.x#90]
This board from Classy Walk hoverboard and has a GD32F130, FORTIOR FD6287T GA1015 and a NCE01P13K 1PKL2. It's marked with OLSZ 0L20190703-V4.1

## Master
![P_20240825_205058](https://github.com/user-attachments/assets/76f26052-30cc-44be-879d-f1ce33f0a668)
![P_20240825_204702](https://github.com/user-attachments/assets/dbd785fd-9302-46f1-a9fa-1bd0f99777ba)

## Slave
![P_20240825_230627](https://github.com/user-attachments/assets/4810b554-3281-49de-aab5-e027d9c3ed4f)
![P_20240825_230609](https://github.com/user-attachments/assets/a918fdab-c2cb-49d5-b416-269e99cef811)

### Config

The hall order was not detected by autoconfig (see below).  Every combination of HALL_A, HALL_B and HALL_C have been tested with every combination of the detected hall pins, PA1, PC13 and PC14. None of them make the motor turn forwards and backwards using `REMOTE_DUMMY`.

```
#ifndef DEFINES_2_x_H
#define DEFINES_2_x_H


//#define SINGLE
//#define MASTER_OR_SINGLE
	
#ifdef MASTER_OR_SINGLE		// layout 2.2 and 2.7 have buzzer on the slave board.
	#define HAS_BUZZER
#endif

// add autodetect #defines below and rename to defines_2-xy.h
// then add another
//	#elif LAYOUT == xy
//		#include "defines_2-xy.h"		// https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/issues/40
// to defines.h and set #define LAYOUT xy in config.h in your chosen GD32E230/MM32SPIN05/GD32F103/GD32F130 target section

// autodetect 2.13 as example only !!!!!!!!!!!!!!!!!!!!!!! begin

// PC13, PC14, PA1
// PC13, PA1, PC14 // Spins smooth one way
// PC14, PC13, PA1
// PC14, PA1, PC13
// PA1, PC13, PC14
// PA1, PC14, PC13

#define HALL_A          PC13
#define HALL_B          PA1
#define HALL_C          PC14

//#define PHASE_A		PB0 // This was detected
//#define PHASE_B		PB1 // This was not detected
//#define PHASE_C		P_UNKOWN

#define LED_RED         PB4
//#define LED_ORANGE            P??
#define LED_GREEN               PB3
#define UPPER_LED               PB2
#define LOWER_LED               PB5
//#define ONBOARD_LED           P??
#define BUZZER          PB9

#define VBATT           PA4
#define CURRENT_DC              PA0 
//#define SELF_HOLD		PB2
//#define BUTTON		PC15
// autodetect 2.13 as example only !!!!!!!!!!!!!!!!!!!!!!! end

// Brushless Control DC (BLDC) defines
#define BLDC_GH PA10		// green	, Tommyboi2001 all bldc pins same as 2.0
#define BLDC_GL PB15		
#define BLDC_BH PA9			// blue
#define BLDC_BL PB14		
#define BLDC_YH PA8			// yellow
#define BLDC_YL PB13		
#define TIMER_BLDC_PULLUP	GPIO_PUPD_NONE	// robo: not sure if some boards indeed nned GPIO_PUPD_PULLUP like 2.2 or 2.3


// Timer BLDC short circuit emergency shutoff define
//#define TIMER_BLDC_EMERGENCY_SHUTDOWN	P_UNKOWN

// choose the serial ports available (and set them to Remote or MasterSlave below)

//#define HAS_USART0	// tx=PB6,rx=PB7	uncomment to connect via 19200 baud serial
// GD32F130 USART0 TX/RX:	(PA9/PA10)AF1	, (PB6/PB7)AF0 , 	(PA2/PA3)AF1 , (PA14/PA15)AF1 GD32F130x4 only!
#ifdef HAS_USART0
	#define USART0_TX	PB6
	#define USART0_RX	PB7
	
	//#define USART0_MASTERSLAVE		// uncomment if this usart is used for master-slave communication
	#define USART0_REMOTE						// uncomment if this usart is used for optional remote control
#endif


#define HAS_USART1	// tx=PA2,rx=PA3	uncomment to connect via 19200 baud serial
// GD32F130 USART1 GD32F130 TX/RX: (PA14/PA15)AF1 , (PA2,PA3)AF1	, (PA8/PB0)AlternateFunction4
#ifdef HAS_USART1
	#define USART1_TX		PA2
	#define USART1_RX		PA3
	
	#define USART1_MASTERSLAVE		// uncomment if this usart is used for master-slave communication
	//#define USART1_REMOTE				// uncomment if this usart is used for optional remote control
#endif


#endif
```

## Hold detection

https://github.com/user-attachments/assets/07c28e3b-4ee3-4a48-b20d-48aaa9bedfdd

Hold detection freezes after the `release OnOff button` stage. When the power button is released, the board just turns off. 

```
0 now A5 : 1106
0 now A6 : 1106
0 now A7 : 1106
0 now B0 : 1106
0 now B1 : 1106
0 now B2 : 1106
0 now B10 : 1106
0 now A1 : 1106
0 now B12 : 1106
0 now A11 : 1107
0 now F6 : 1106
0 now F7 : 1107
0 now A12 : 1106
0 now A15 : 1107
0 now B3 : 1106
0 now B4 : 1107
0 now B5 : 1106
0 now B6 : 1106
0 now B7 : 1105
0 now B8 : 1107
0 now B9 : 1106
release OnOff button
ENTER to begin, 's' for next stage, 'q' for menu

```

## Hall / Hall order detection

https://github.com/user-attachments/assets/b025cb2a-ecb6-43c4-bc1f-d2cba355bd90

The hall detection works but the hall order detection does not detect the order.

```
hit ENTER:
:-)

:-)
:-)
:-)
:-)
:-)
:-)
:-)autodetect menu:
0: do all
1: VBatt
2: Hold
4: Led
5: Hall
8: Results

hall and phase currents: 'r'=restart	's'=save
try 0 = C13	HALL
try 1 = C14	HALL
try 2 = C15	
try 3 = F0	
try 4 = F1	
try 5 = A0	
try 6 = B11	
try 7 = A4	
try 8 = A5	
try 9 = A6	

```
## Slave board

The slave board flashes but does not appear to turn on. There is obviously no switch to bridge. 

### Flashing

The master board can't be flashed when being powered via the bench power supply but can be flashed when powered via the 3.3V line on the ST-LINK flasher. I read elsewhere that this is bad and the ST-LINK does sometimes behave strangely. (Maybe too much current is being drawn?) Any idea why it does not flash while being powered by the bench power supply and are there any workarounds? Maybe it would be better to have an external 3.3V power supply rather than using the 25.9V bench power supply?

### Version number

Any idea how this board would fit into the existing version number scheme? It says 0L20190703-V4.1 on it.

**Comment by RoboDurden** [comment]
Please post the full autodetect log.
And from your video i see that you do not use an arduino with my autodetect.ino as the serial log data proxy.
Only my autodetect.ino can save the tested HOLD pin so on repowering the board, the autodetect firmware can retrieve the last tested button from autodetect.ino and reconize this as the HOLD button..

You should also to run the normal autodetect procedure with "0". Then i think the first pin to detect is the voltage adc VBATT.

And anyway, if the HOLD detection does not work, then simply keep the onOff button bridged and skip this detection stage.
So we see if the different led and  the buzzer can be detected.

The hall dectection normaly needs the VBATT, because the pwm ratio must be adjusted for 26V or 36V...

**Comment by RoboDurden** [comment]
I can not read the log ouput on your hall video.
Does the autodetect indeed finds all three hall pins ?
I only see two pins detected in the beginning, then you mention a phase being detected. But the scan continous until all possible pins have been tested but no third hall found ?

**Comment by I-hate-2FA** [comment]
hall detect work normally 3 pin found but no order is valid
self hold and button is not required for the firmware to function so you dont need to test those

**Comment by hoverboardhavoc** [comment]
Thanks for the help with this!

I'm running `autodetect.ino` on a [Blue Pill](https://stm32-base.org/boards/STM32F103C8T6-Blue-Pill.html). I can send pull request with the changes I made to make it work. 

Yes, all three hall pins are detected. I failed to paste the logs properly because the buffer is weird in the Aurdino IDE. Creating an entire log has proved problematic, If I run the whole detect routine using, 0 the board just dies when I unbridge the power button. If I detect VBATT (PA4) prior to detecting the hall sensors, hall order detection still fails. The three HALL pins are detected, `PC13, PC14, PA1`. 

Every combination of HALL_A, HALL_B and HALL_C have been tested with every combination of the detected hall pins, PA1, PC13 and PC14. None of them make the motor turn forwards and backwards using REMOTE_DUMMY.

**Comment by RoboDurden** [comment]
Yes, he could already proceed with a RemoteDummy firmware and simply test the 6 hall permutation with six different firmwares.
But i would autodetect the rgb leds first so that they will blink according to the hall sensors:

```
		#ifdef REMOTE_DUMMY
			#define TEST_HALL2LED	// led the 3-led panel blink according to the hall sensors

```

**Comment by hoverboardhavoc** [comment]
I only have two color leds to play with unfortunately. There is only green and red (no orange). But they do light up when I manual rotate the wheel when using `REMOTE_DUMMY`

**Comment by I-hate-2FA** [comment]
you should use putty as the default serial terminal if you are using windows

**Comment by RoboDurden** [comment]
If i remember correctly, my HOLD detection has two stages. First it tries to detect the power fail quickly enough to keep code running. If this fails it will use the autodetect.ino to save the last pin tested and recover that information after repowering.

As your board seem to turn off immediately, setting all possible pins to INPUT_PULLUP does not seem to activate the HOLD button.
maybe OUTPUT and digitalWrite(1) is needed. or even digitalWrite(0).

I think you should try to detect as many other pins as possible, so when everything else runs, you have less to try-and-error with the HOLD/BUTTON pins.

Yes, if you have detected UP or DOWN, you can manualy define the orange pins as such a one.

If the led blinking does confirm that you have found the correct three hall pins, than i think that one of the six possible permutations must do.

Because my hall pin detection stage does make the motor spin. So the the activations of the six mosfet seem to work as they are supposed to.

My autodetect log output does support the Arduino serial terminal.

**Comment by I-hate-2FA** [comment]
> I only have two color leds to play with unfortunately. There is only green and red (no orange). But they do light up when I manual rotate the wheel when using `REMOTE_DUMMY`. Maybe I could try with the up or down led as the third led

yes this is what i do when i develop on the spin06 board where there is only 2 led driver on master
your board seems to contain 7 led driver transistor though which is quite allot

**Comment by hoverboardhavoc** [comment]
I tried using red, green and up and I think this indicates that I have correctly detected the hall pins. I have tried all the possible different orders and nothing worked but I can try again. 

https://github.com/user-attachments/assets/4519ee5b-bdbb-4e15-9145-48e43429ce4d

**Comment by I-hate-2FA** [comment]
in fact i probably know why it doesnt work

**Comment by hoverboardhavoc** [comment]
Am I OK to flash the board using the power from the ST-LINK? I can't seem to flash the board when it's powered from the PSU.

**Comment by hoverboardhavoc** [comment]
@I-hate-2FA The serial is running on a mac currently. I can try putty on Windows. Thanks

**Comment by I-hate-2FA** [comment]
this is gen2, you should always use the stlink to power the board when flashing, gen1 dev dont know what they are doing

**Comment by I-hate-2FA** [comment]
1. make sure the battery low protection is disabled
2. make sure invert lowside is set correctly

**Comment by RoboDurden** [comment]
I do not think that he needs to change "invert lowside" 

```
	// Set up the output channel parameter struct
	timerBldc_oc_parameter_struct.ocpolarity 		= TIMER_OC_POLARITY_HIGH; //HIGH: CHx_O is the same as OxCPRE , LOW: CHx_O is contrary to OxCPRE
	timerBldc_oc_parameter_struct.ocnpolarity 	= TIMER_OCN_POLARITY_LOW; //HIGH: CHx_ON is contrary to OxCPRE, LOW: CHx_O is the same as OxCPRE
```

As my hall detection stage is already able to make the motors spin.

**Comment by RoboDurden** [comment]
No need for putty. He obviously can send comands to autodetect.bin and receive the log output :-)

**Comment by hoverboardhavoc** [comment]
Success!! It seems really smooth :)

@ I-hate-2FA hinted at how voltage problems and `BAT_CELLS` was set incorrectly. Its defined both on line 14 and 36 in `config.h`

https://github.com/user-attachments/assets/1842cbcf-16f3-4d18-8af1-3d909ffa41b9

Thanks so much for the help.

So the remaining issues I have are:
* Making the slave board power up
* Hold / power button detection

**Comment by hoverboardhavoc** [comment]
If I send a PR for this board, what should I call it?

**Comment by RoboDurden** [comment]
The slave usualy gets 5V or 12V via the Master-Slave cable which makes it wake up.
So simply measure what positive voltage you have on the master slave cable, and feed this to the correct pin of the slave board.

**Comment by RoboDurden** [comment]
Your defines you be named `defines_2-1-20.h`

**Comment by hoverboardhavoc** [comment]
Is there any lower bound to the voltage I can use. Can I try 12V? I guess it will be slow if it works?

**Comment by RoboDurden** [comment]
Concerning the HOLD button.. I guess i would add to main.c

```
		#define COUNT_PinDigital 31
		PinAD aoPin[COUNT_PinDigital] = {
				{PC13,"C13",0}	,{PC14,"C14",0}	,{PC15,"C15",0}	,{PF0,"F0",0}	,{PF1,"F1",0}		,
				{PA0,"A0",1}		,{PB11,"B11",0}		,{PA4,"A4",1}		,{PA5,"A5",1}	,{PA6,"A6",1} 	,
				{PA7,"A7",1}		,{PB0,"B0",1}		,{PB1,"B1",1}		,{PB2,"B2",0}	,{PB10,"B10",0}	,
				{PA1,"A1",1}	,{PB12,"B12",0}	,{PA11,"A11",0}	,{PF6,"F6",0}	,{PF7,"F7",0} 	,
				{PA12,"A12",0}	,{PA15,"A15",0}	,{PB3,"B3",0}		,{PB4,"B4",0}	,{PB5,"B5",0}	,
				{PB6,"B6",0}		,{PB7,"B7",0}		,{PB8,"B8",0}		,{PB9,"B9",0}	,{PA2,"A2",1}	,
				{PA3,"A3",1}
				};
```

with all the pins removed that you already know to have a different function. 
I you had an autodetect.bin with PA2 and PA3 as rx tx pins.

Then before the while(1) loop add something like

```

		for (i=0;i<COUNT_PinDigital; i++)	
		{
				//pinModePull(aoPin[i].i,GPIO_MODE_INPUT,GPIO_PUPD_PULLUP);
				pinMode(aoPin[i].i,GPIO_MODE_OUTPUT);
				digitalWrite(aoPin[iTest].i,1);	// every remaining io pin might be the SELF_HOLD pin
		}
```

And then release the onoff button when the dummy starts spinning.
That way you might be able to find out, what make the HOLD transistor being active.

**Comment by I-hate-2FA** [comment]
> Is there any lower bound to the voltage I can use. Can I try 12V? I guess it will be slow if it works?

15v

**Comment by hoverboardhavoc** [comment]
> Concerning the HOLD button.. I guess i would add to main.c
> 
> ```
> typedef struct {
>    uint32_t i;
>    char*  s;
>    uint8_t  wState;   // 1=adc
>    int16_t  iAdc;  
> } PinAD;
> 
> 
> 		#define COUNT_PinDigital 31
> 		PinAD aoPin[COUNT_PinDigital] = {
> 				{PC13,"C13",0}	,{PC14,"C14",0}	,{PC15,"C15",0}	,{PF0,"F0",0}	,{PF1,"F1",0}		,
> 				{PA0,"A0",1}		,{PB11,"B11",0}		,{PA4,"A4",1}		,{PA5,"A5",1}	,{PA6,"A6",1} 	,
> 				{PA7,"A7",1}		,{PB0,"B0",1}		,{PB1,"B1",1}		,{PB2,"B2",0}	,{PB10,"B10",0}	,
> 				{PA1,"A1",1}	,{PB12,"B12",0}	,{PA11,"A11",0}	,{PF6,"F6",0}	,{PF7,"F7",0} 	,
> 				{PA12,"A12",0}	,{PA15,"A15",0}	,{PB3,"B3",0}		,{PB4,"B4",0}	,{PB5,"B5",0}	,
> 				{PB6,"B6",0}		,{PB7,"B7",0}		,{PB8,"B8",0}		,{PB9,"B9",0}	,{PA2,"A2",1}	,
> 				{PA3,"A3",1}
> 				};
> ```
> 
> with all the pins removed that you already know to have a different function. I you had an autodetect.bin with PA2 and PA3 as rx tx pins.
> 
> Then before the while(1) loop add something like
> 
> ```
> 
> 		for (i=0;i<COUNT_PinDigital; i++)	
> 		{
> 				//pinModePull(aoPin[i].i,GPIO_MODE_INPUT,GPIO_PUPD_PULLUP);
> 				pinMode(aoPin[i].i,GPIO_MODE_OUTPUT);
> 				digitalWrite(aoPin[iTest].i,1);	// every remaining io pin might be the SELF_HOLD pin
> 		}
> ```
> 
> And then release the onoff button when the dummy starts spinning. That way you might be able to find out, what make the HOLD transistor being active.

Thanks, using this method I discovered that hold is PB12

**Comment by RoboDurden** [comment]
So it does need output and digitalWrite(PB12,1) to activate the HOLD transistor ?

**Comment by I-hate-2FA** [comment]
it is better to connect the button pin directly to your external micro controller and disable self hold, so you can control power to the board on demand(assume the mcu is powered externally)

**Comment by I-hate-2FA** [comment]
> @I-hate-2FA , what transistor/optocoupler do you suggest to emulate the onOff button ?

you do not need one, just connect the button connector directly to esp gpio, the transistor is built in the hoverboard, if they have common gnd itll work

**Comment by RoboDurden** [comment]
?? the 2pin button (onOff) socket has 36V on one pin ?

**Comment by I-hate-2FA** [comment]
connect gpio to the one that dont have 36v, 36v one is not needed

**Comment by RoboDurden** [comment]
yes that might work. But if the other pin is connected to ESP32, it will kill the ESP.
But yes, if people will use a ESP32  for uart control, they might be better of turn the hoverboard on off from there.
On the other hand, the hoverboard could be used to power the esp32.

**Comment by I-hate-2FA** [comment]
i already tried it irl it worked on my board

**Comment by hoverboardhavoc** [comment]
> Now in RemoteAutodetect.c on line 365 you could change
> 
> ```
> 		pinModePull(aiPinScan[SCAN_SELF_HOLD],GPIO_MODE_INPUT,GPIO_PUPD_PULLUP);
> 		//pinMode(aiPinScan[SCAN_SELF_HOLD],GPIO_MODE_OUTPUT);
> 		//digitalWrite(aiPinScan[SCAN_SELF_HOLD],1);	// SELF_HOLD pin to high
> ```
> 
> to
> 
> ```
> 		aiPinScan[SCAN_SELF_HOLD] = PB12;
> 		HidePinDigital(PB12);
> 		//pinModePull(aiPinScan[SCAN_SELF_HOLD],GPIO_MODE_INPUT,GPIO_PUPD_PULLUP);
> 		pinMode(aiPinScan[SCAN_SELF_HOLD],GPIO_MODE_OUTPUT);
> 		digitalWrite(aiPinScan[SCAN_SELF_HOLD],1);	// SELF_HOLD pin to high
> ```
> 
> Then you might be able to release the onOff button after you have choosen the BUTTON detect stage and that way discover the button input

Having made this change, the autoconfig just seems to go into a loop when doing `2: Hold`. 3 is not available in the menu, but I can type it anyway but it also goes into an infinite loop:

```
release OnOff button
ENTER to begin, 's' for next stage, 'q' for menu
 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15 F0 F1 A0 B11 A5 A6 A7 B1 B10 A11 F6 F7 A12 A15 B6 B7 B8 B9 C15autodetect menu:
0: do all
1: VBatt
2: Hold
4: Led
5: Hall
6: HallOrder
7: CurrentDc
8: Results

```

**Comment by I-hate-2FA** [comment]
and itll only make sense if you power the esp externally yes

**Comment by hoverboardhavoc** [comment]
Also, just out of curiosity, why was the autodetect not able to do the `digitalWrite(aiPinScan[SCAN_SELF_HOLD],1);` Was I just using it wrong somehow?

**Comment by I-hate-2FA** [comment]
because pinmode output in autodetect is not allowed

**Comment by RoboDurden** [comment]
As you can see this OUTPUT + digitalWrite was the way i did it - before @I-hate-2FA  convinced me that this might create a short cut on some boards. So i switched to the INPUT_PULLUP method.

Yes, you can add `wMenuStage |= AUTODETECT_Stage_Button;` to make this stage visible in the menu.

**Comment by RoboDurden** [comment]
If i remember correctly this button detect stage also has two parts, one with onoff bridged and one with releasing it. So the code can detect the difference and find the pin.

**Comment by hoverboardhavoc** [comment]
I figured out the last piece of the puzzel for the master board, `BUTTON`:

````
#ifndef DEFINES_2_x_H
#define DEFINES_2_x_H

#define HALL_A PC13
#define HALL_B PA1
#define HALL_C PC14
#define LED_RED PB4
#define LED_GREEN PB3
#define UPPER_LED PB2
#define LOWER_LED PB5
#define BUZZER PB9
#define VBATT PA4
#define CURRENT_DC PA0 
#define SELF_HOLD	PB12
#define BUTTON PA12

// Brushless Control DC (BLDC) defines
#define BLDC_GH PA10		// green	, Tommyboi2001 all bldc pins same as 2.0
#define BLDC_GL PB15		
#define BLDC_BH PA9			// blue
#define BLDC_BL PB14		
#define BLDC_YH PA8			// yellow
#define BLDC_YL PB13		
#define TIMER_BLDC_PULLUP	GPIO_PUPD_NONE	// robo: not sure if some boards indeed nned GPIO_PUPD_PULLUP like 2.2 or 2.3

#define HAS_USART1	// tx=PA2,rx=PA3	uncomment to connect via 19200 baud serial
// GD32F130 USART1 GD32F130 TX/RX: (PA14/PA15)AF1 , (PA2,PA3)AF1	, (PA8/PB0)AlternateFunction4
#ifdef HAS_USART1
	#define USART1_TX		PA2
	#define USART1_RX		PA3
	
	#define USART1_MASTERSLAVE		// uncomment if this usart is used for master-slave communication
#endif

#endif
```

I'll send a PR in the coming days

**Comment by hoverboardhavoc** [comment]
I attempted to make a PR but it says I have an "entirely different commit history"

https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/compare/main...hoverboardhavoc:hoverboardhavoc/classy-walk?expand=1

**Comment by I-hate-2FA** [comment]
no because then you will not be able to download the code in browser

**Comment by I-hate-2FA** [comment]
all big project issue tracker and wiki have seperate repo like esphome

**Comment by RoboDurden** [comment]
Yes, This home repo had become to big in size. The GD32 code repo is easy to download.

**Comment by RoboDurden** [comment]
Thanks for the pull request. I improved your missing orange led code in bldc.c :

```
	#ifdef TEST_HALL2LED
		#ifdef LED_ORANGE
			digitalWrite(LED_ORANGE,hall_b);
		#elif defined(UPPER_LED)
			digitalWrite(UPPER_LED,hall_b);
		#elif defined(LOWER_LED)
			digitalWrite(LOWER_LED,hall_b);
		#else
			if (hall_b)
			{
				digitalWrite(LED_GREEN,(steerCounter%2) < 1);
				digitalWrite(LED_RED,(steerCounter%2) < 1);
			}
		#endif
		digitalWrite(LED_GREEN,hall_a);
		digitalWrite(LED_RED,hall_c);
	#endif
```

**Comment by RoboDurden** [comment]
> or another USART but no connector. Maybe I'll need to solder something on if I want to connect the board to a micro controller and a slave



> I think there are pads for another USART but no connector. Maybe I'll need to solder something on if I want to connect the board to a micro controller and a slave

Yes, the second uart port (usually PB6 and PB7 ) would be good to have master-slave communication.
But i added the RemoteUartBus with which you can control two or more single boards directly from ESP32 (or likewise).

What empty uart header are you reffering to ?
Please show it on a photo.
I see some bluetooth modules on your photos.
Sometimes they only offer a TX or RX line and not both..

**Comment by RoboDurden** [comment]
The slave board has a red 4 pin header mounted. That might be three additional led outputs.

**Comment by I-hate-2FA** [comment]
it is 2024 we should always the the same uartbus protocol

**Comment by hoverboardhavoc** [comment]
> > I think there are pads for another USART but no connector. Maybe I'll need to solder something on if I want to connect the board to a micro controller and a slave
> 
> Yes, the second uart port (usually PB6 and PB7 ) would be good to have master-slave communication. But i added the RemoteUartBus with which you can control two or more single boards directly from ESP32 (or likewise).
> 
> What empty uart header are you reffering to ? Please show it on a photo. I see some bluetooth modules on your photos. Sometimes they only offer a TX or RX line and not both..

I put a red ring around what I thought might be a UART
![361253441-dbd785fd-9302-46f1-a9fa-1bd0f99777ba](https://github.com/user-attachments/assets/314c1515-47f2-4ef9-ab83-35b55c71119a)

**Comment by RoboDurden** [comment]
but i see three little smd transistors above this 4pin header, that is why we think it is another led driver port.

Try to trace the PB6 and PB7 of the mcu. I guess at least one of those will lead to the bluetooth module.
If you are lucky, both rx and tx will end there. Then you have the second uart.

**Comment by I-hate-2FA** [comment]
there is 3 pin connector near antenna its highly likely

**Comment by DDoru** [comment]
Hello, i have couple of this boards and will be nice to use them. I read and read and read again the wiki and i just don't understand where is the bin file for this board to flash....there is any tutorial to guide me? In folder BinariesReadyToFlash there is no hoverboard 2.1.20.bin file.... please if possible some help. I have more hoverboard boards if is any help, but this type of board i have 6-7 pairs.

**Comment by RoboDurden** [comment]
If have uploaded three binaries for you: https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/tree/main/BinariesToTest

Please start with the dummy, which should spin motor backwards and forwards.
Your board only has one uart :-/

Tell me how it went, then i can move the binaries to /readytoflash.

Only test with a 2A constant current power supply !
Or the 1.5A charger :-)

**Comment by RoboDurden** [comment]
read the wiki here for all the different methods to flash firmware.
If you have a windows machine, please rename binary into hoverboard.bin and use autoflash.bat

I have just make the project compile successfully with visualCode+platformIO ! If you have that installed, please compile and upload the firmware yourself.

Installing Keil ide is also easy.

**Comment by DDoru** [comment]
You are a genius, autoflash.bat  worked perfect. Maybe i need to change something in configurations....i will install visualCode+platformIO and see if I can do something to work.
For the moment only the motor with power button spin for couple of seconds and then stop.
Thank you for your time! 

https://github.com/user-attachments/assets/36845fe7-6c3e-4026-aec5-8a1c1236cdd5

https://github.com/user-attachments/assets/02a6356e-0bd3-42f8-9918-dee05fee27c9

**Comment by DDoru** [comment]
I need to install hoverboard 2.1.20 single UartBus id0 and id1 and then connect with sep32 and test serial correct?

**Comment by RoboDurden** [comment]
I can confirm this over current shutdown with my 6A power supply set to 4A. The 1.5A charger makes no problem !!

**Comment by RoboDurden** [comment]
the shutoff function gets called ! if i disable it, there is still a speed=0 stop (which probably causes the overcurrent when happening with motor spinning), but the board stays alive:

**Comment by RoboDurden** [comment]
It is the VBATT pin being wrongly defined in the definesXY.h file.
wrong battery reading -> shutoff.
Strange that ShutOff doesnt work when connected to the charger.

**Comment by DDoru** [comment]
The cable from slave is not PB6 and PB7 as is say in the arduino sketch,  slave cable is PA2 and PA3 .....
I have just the master board connected, esp32 i changed to Serial2 because with serial1 did not work not even the serial monitor.
I think I touch something that the first esp was burnt.
I use the Testspeed.ide.... esp connected with GND Rx Tx , also I switch Rx Tx wire.....the same.  Will see...
I also try on the Rx Tx for the slave board and stil nothing


https://github.com/user-attachments/assets/54e7eb14-5fbb-413f-97ae-b6bc2a11ad00

**Comment by RoboDurden** [comment]
The PB6/PB7 was not listed in the defines_2-1-20.h file, therefore i did only compile the RemotUartBus, so that you can connect both boards in parallel via the PA2/PA3 master/slave header.
If it is indeed possible to solder to PB6 AND PB7, please give a photo of the final result. Then you could use RemoteUart and connect the boards with the masterslave cable.

I am still working on the VBATT issue. my pin PA4 was correct with my 2.1.11 test setup as i did the autodetect again (the detection of latch and button did work without the bord shutting off). so there might be some bug that wrecks vbatt reading in the normal firmware, which then causes the shutdown.

**Comment by I-hate-2FA** [comment]
very likely adc control

**Comment by RoboDurden** [comment]
Still there ? I can upload new binaries for you with that PA4 pin removed, then your motor would not stop.

**Comment by DDoru** [comment]
The boards and motor are in other place, i will be there Monday or Tuesday and will try again

**Comment by RoboDurden** [comment]
Mhhhhhhhhhhhhhhh, there is nothing wrong with VBATT pin :-(
I simply only had a few hours sleep last night.
Of course, if we test with 25 V but the binaries expect 10s 36V, then the board will turn off.

Anyway, i have recoded the battery level led output and added a soft stop of the motor when shutoff :-)

**Comment by RoboDurden** [comment]
@DDoru  what is tx and what rx ? blue and yellow cable

**Comment by RoboDurden** [comment]
i have uploaded new binaries to test, now 7s for 25V battery/powerSupply

**Comment by DDoru** [comment]
> [@DDoru](https://github.com/DDoru) what is tx and what rx ? blue and yellow cable

yes

**Comment by I-hate-2FA** [comment]
please disable battery protection by default @RoboDurden , i have much of same problem

**Comment by I-hate-2FA** [comment]
allot of people come in support and say everything is working in pinfinder, but in main the motor doesnt spin

**Comment by I-hate-2FA** [comment]
> maybe to remove IMU

no need to remove imu, you can leave everything in place or remove the series resistor from imu to gd32

**Comment by I-hate-2FA** [comment]
i have tested on another board

**Comment by DDoru** [comment]
> > maybe to remove IMU
> 
> no need to remove imu, you can leave everything in place or remove the series resistor from imu to gd32

the two resistors there in the pictures are going to the GND they are not in series, but if works with serial from slave board...there is no need for that as RoboDurden say

**Comment by DDoru** [comment]
> yes its better to use uartbus it is the best thing ever a standardized protocol finally

sorry i don't understand , is better to use PB6 and PB7 from IMU or better to use Pa2 and PA3 from slave board?

**Comment by I-hate-2FA** [comment]
it doesnt matter which serial you use but is better to use uartbus protocol which only required 1 uart per board and it can be either but obviousely pa2 pa3 is more convinient

**Comment by RoboDurden** [comment]
Mhhh, @I-hate-2FA is the Linux believer who thinks that everyone should read his fucking manuals ;-)

UartBus is more error prone for beginners:
1. you do not get serial feedback when the connection from esp to hover is broken
2. you can not make use of the nice master-slave cable that comes with the hoverboard
3. you might need a diodes to make the hover->esp work, and BDCs will get the diode directon wrong
4. you can get the multiple uartBus ids wrong.
5. you have to take care of speed/steer yourself in esp code.
6. etc.

So  when your master board has the two uarts and you only want to drive two motors anyway, OF COURSE you should NOT use RemoteUartbus but the simpler RemoteUart

**Comment by DDoru** [comment]
This is secondary serial for this board, PB6 (Tx)  and PB7 (RX), is connected to two resistors pull up to 3,3v and then to  the pads of other IMU model, not populated.

![Image](https://github.com/user-attachments/assets/4801b594-f1d2-4f06-86e3-ee33329f0c34)

![Image](https://github.com/user-attachments/assets/69ba1f7a-7f4d-405d-8c1f-b707cbb7ee06)

**Comment by RoboDurden** [comment]
Those little smd resistors to the left are pull-ups to 3.3V ?
And the populated IMU does not connect to PB6/PB7 as well ?

> hoverboard 2.1.20 master Dummy tested from 20 to 42V spinning 2-3 second ant then it stop the motor.

You mean that 43V made the motor stop ?
Or does the motor always stop after 2-3 seconds. In the video, it is spinng for at least 7 seconds.

**Comment by RoboDurden** [comment]
but the motor behavior in the video is exactly as it should be with RemoteDummy: spinning forward and backwards.
If the populated IMU is also connected to PB6/PB7 this might spoil uart communication. But IMU mostly use I2C-slave, which will stay silent when not recieving a propper i2c request :-)

**Comment by DDoru** [comment]
![Image](https://github.com/user-attachments/assets/dd405f1a-0d1d-480e-9de0-7abc7853ca87)

I mounted diodes like this and try to spin both motors.....not working, What I am doing wrong?

 I am using hoverboard 2.1.20 single UartBus id0.bin for master and hoverboard 2.1.20 single UartBus id1.bin for slave, also i test 0 for both boards...or 1 for both boards, same result only master motor spin.

 I check 12v 5v and 3.3v on slave board...all are ok.

**Comment by DDoru** [comment]
I used the binary you sent, I did not try the platformIo yet, just try to control both motor first...and later will see.

![Image](https://github.com/user-attachments/assets/9408a787-302e-4575-82e0-ca1ab31d1bf5)

**Comment by DDoru** [comment]
The slave board never work, no matter what Id i flash.

Will install tomorrow keil or platformIO ( what do you recommend for this split boards) and I will try to compile myself, I don't know what you compile and I don't want to bother you too much :))). Thank you!  For sure I will come back with setup for compiling.

**Comment by RoboDurden** [comment]
check the new binaries i uploaded.
Yes, the old ones might not work on the slave because it has not onOff button to detect.
Choose the new id1_DisableButton

**Comment by DDoru** [comment]
> check the new binaries i uploaded. Yes, the old ones might not work on the slave because it has not onOff button to detect. Choose the new id1_DisableButton

Thank you! will test new binaries tomorrow morning.

**Comment by RoboDurden** [comment]
For slave board you should uncomment 

`		//#define DISABLE_BUTTON	// this is the opposite of former CHECK_BUTTON define.
`

Maybe reduce BAT_CELLS to 7 if you test with 25V. Then battery buzzer/led output will be correct.
And start with BLDC_BC as this is simpler motor drive.
SINGLE+RemoteDummy should make the motor spin forwards and backwards.

**Comment by RoboDurden** [comment]
I have just checked RemoteUart with my Gen2.1.4 and is working nicely with my ESP S2 mini.
Make sure you have selected the usart that you are actually connecting to in config:

```c++
	#if defined(REMOTE_UART) || defined(REMOTE_UARTBUS) || defined(REMOTE_CRSF)
		#define REMOTE_USART				1 	// 	1 is usually PA2/PA3 and the original master-slave 4pin header
																		//	0 is usually PB6/PB7 and the empty header close to the flash-header
																		//	2 is usually PB10/PB11 on stm32f103 boards
	#endif
```

**Comment by DDoru** [comment]
> I have just checked RemoteUart with my Gen2.1.4 and is working nicely with my ESP S2 mini. Make sure you have selected the usart that you are actually connecting to in config:
> 
> 	#if defined(REMOTE_UART) || defined(REMOTE_UARTBUS) || defined(REMOTE_CRSF)
> 		#define REMOTE_USART				1 	// 	1 is usually PA2/PA3 and the original master-slave 4pin header
> 									//	0 is usually PB6/PB7 and the empty header close to the flash-header
> 									//	2 is usually PB10/PB11 on stm32f103 boards
> 	#endif

I use #define REMOTE_UARTBUS, was working when you compile, if i flash yours now is working. When I compile ...nothing just different sound in buzzer.

**Comment by RoboDurden** [comment]
If have update my repo and tested RemoteUart again on PA2/Pa3 with Gen2.1.4, please update your local repo.
You should start with RemoteUart and a single board.
RemoteUartBus is a bit more error prone to setup.

Please try to write full english sentences! Sometimes it is hard to understand what you  are writing.

**Comment by DDoru** [comment]
> If have update my repo and tested RemoteUart again on PA2/Pa3 with Gen2.1.4, please update your local repo. You should start with RemoteUart and a single board. RemoteUartBus is a bit more error prone to setup.
> 
> Please try to write full english sentences! Sometimes it is hard to understand what you are writing.

Ok, will test in the morning RemoteUart on PA2/PA3. Thank you! Sorry about my English, is not my mother language.

**Comment by DDoru** [comment]
Last repositories, configSelect.h not found.
![Image](https://github.com/user-attachments/assets/8de1fcb8-db91-4418-83ae-37acd253deee)

**Comment by DDoru** [comment]
We thank you!

**Comment by DDoru** [comment]
For this board model, the slave board needs 12V bridge from master board ( red wire from input of 5v regulator) because on the communication cable is only 5V. Needs 12V for mosfet driver IC. This was the reason why the slave Board was not spinning the motor.

![Image](https://github.com/user-attachments/assets/a0580f0e-2369-46e3-a009-0e03ac194976)

![Image](https://github.com/user-attachments/assets/96213c33-6062-4116-b3e6-18479274816c)

![Image](https://github.com/user-attachments/assets/f168ae0b-2733-487d-8f37-850369024536)

**Comment by RoboDurden** [comment]
Looks perfect to me. Did you need the diodes for this dual RemoteUartBus setup ?
Sometimes you can simply connect rx and tx in parallel.
And only for more boards like 4-wheeler are the diodes mandatory.

BLDC_SINE did already work yesterday ?

Normally, the +pin of the master-slave cable has +12V or +15V, if the slave needs that.
Or the +5V on the master-slave turns on the 15V dcdc step down converter of the slave.

How were these two boards connected originally ????

**Comment by I-hate-2FA** [comment]
this board is designed as double single , both board have buck converter, but obviousely after flash the circuit need to be modified on slave to make buck converter always active

**Comment by DDoru** [comment]
I connected Rx and TX in parallel no diodes.

BLDC_SINE worked yesterday or two day a go, now does not work.

Original connection was with 5 V GND Rx Tx, 

PB12 is connected to PIN3 of DC DC converter through a resistor, maybe that is the start of DC DC converter.

DC DC converter SP6504 8 pin.....no datasheet available.

**Comment by RoboDurden** [comment]
new update online.

**Comment by RoboDurden** [comment]
configDebug.h is only for me and your configSelect.h should look like

```c++
// this file will not be updated to github :-)

//#define CONFIG_DEBUG	// so robo can work on configDebug.h

#ifdef CONFIG_DEBUG
	#include "../Inc/configDebug.h"
#else
	#include "../Inc/config.h"
#endif
```

**Comment by RoboDurden** [comment]
your error indicates that you have wrongly set remote uart to 2
```c++
	#if defined(REMOTE_UART) || defined(REMOTE_UARTBUS) || defined(REMOTE_CRSF)
		#define REMOTE_USART		0 	// 	1 is usually PA2/PA3 and the original master-slave 4pin header
				//	0 is usually PB6/PB7 and the empty header close to the flash-header
				//	2 is usually PB10/PB11 on stm32f103 boards
	#endif

```

**Comment by DDoru** [comment]
> Don't know why the motors block at the end of the video.

No worry, the motors are working ok no problem I modify the code for ESP32 to adjust the speed forward and reverse by pressing "u" and "d" on keyboard. 

The only strange thing ( but no problems ) is that at speed 100 the motor sound like BLDC not SINE, some kind of resonance at that speed.


https://github.com/user-attachments/assets/e74688f7-837d-4eb0-8df7-07714df7520e

**Comment by DDoru** [comment]
//#define INACTIVITY_TIMEOUT 	8        	

How can i disable inactivity? If I comment out i have an error

`Build started: Project: Hoverboard
*** Using Compiler 'V6.23', folder: 'C:\Users\User\AppData\Local\Keil_v5\ARM\ARMCLANG\Bin'
Build target 'GD32F130'
Src/main.c(340): error: use of undeclared identifier 'INACTIVITY_TIMEOUT'
  340 |                         if (inactivity_timeout_counter > (INACTIVITY_TIMEOUT * 60 * 1000) / (DELAY_IN_MAIN_LOOP + 1))
      |                                                           ^
1 error generated.
compiling main.c...
".\Objects\Hoverboard.axf" - 1 Error(s), 0 Warning(s).
Target not created.
Build Time Elapsed:  00:00:01`

**Comment by RoboDurden** [comment]
in bldcSINE.c , i disable my angle interpolation for speed less than 100. Also for low speeds (= small `safe_hall_time_step` )

```c++
	if (ABS(pwmGo) > 100)
	{
		if (sector != sectorLast) 
		{
			sectorChange = (sector - sectorLast + 6) % 6;	// normaly +1 or -1 but could be more if a hall position is skipped because of PWM_FREQ sampling frequency
			if (sectorChange > 3) sectorChange -= 6;  // Handle reverse direction
		}
		
		if (	(safe_hall_time_step > 0) && (safe_hall_time_step < (PWM_FREQ/20))	)	// >50ms for one hall change is to slow to interpolate
		{
			uint16_t iDT = buzzerTimer>safe_hall_time_last ? buzzerTimer-safe_hall_time_last : buzzerTimer + (0x00010000-safe_hall_time_last);
			
			fDT = (float)iDT /safe_hall_time_step;		// at constant speed, fDT == 1 would be right before a new sector gets detected by the hall inputs
			if (fDT < 1.5)	// not for heavy breaking
				angle_deg_add = sectorChange * (60.0f  * fDT) ;
		}
		int16_t iOffset = sectorChange < 0 ? 390 : 330;		// +-30° +360° to prevent negative angle
		angle_deg_final  = angle_deg + angle_deg_add + iOffset;
	}
	else
		angle_deg_final = angle_deg;

```

**Comment by RoboDurden** [comment]
But currently i do not add the 30° iOffset to put the angle in the middle of the 60° sectors for speeds below 100.
Please give this code a try and report if motor behavior is better for speed below 100 :-)

```c++
	if (sector != sectorLast) 
	{
		sectorChange = (sector - sectorLast + 6) % 6;	// normaly +1 or -1 but could be more if a hall position is skipped because of PWM_FREQ sampling frequency
		if (sectorChange > 3) sectorChange -= 6;  // Handle reverse direction
	}
	int16_t iOffset = sectorChange < 0 ? 390 : 330;		// +-30° +360° to prevent negative angle
	
	if (ABS(pwmGo) > 100)
	{
		if (	(safe_hall_time_step > 0) && (safe_hall_time_step < (PWM_FREQ/20))	)	// >50ms for one hall change is to slow to interpolate
		{
			uint16_t iDT = buzzerTimer>safe_hall_time_last ? buzzerTimer-safe_hall_time_last : buzzerTimer + (0x00010000-safe_hall_time_last);
			
			fDT = (float)iDT /safe_hall_time_step;		// at constant speed, fDT == 1 would be right before a new sector gets detected by the hall inputs
			if (fDT < 1.5)	// not for heavy breaking
				angle_deg_add = sectorChange * (60.0f  * fDT) ;
		}
	}
	angle_deg_final  = angle_deg + angle_deg_add + iOffset;

```

**Comment by DDoru** [comment]
Ok will test tomorrow morning, thank you!

**Comment by RoboDurden** [comment]
Not it got worse.
This might be better than what is currently only:

**Comment by DDoru** [comment]
This video is before modification, slow speed 50 high speed 100 noise at 100 speed. Under 50 the motor does not spin ( does not matter)

https://github.com/user-attachments/assets/70571f61-ed51-4bc7-9ace-cbdd7d48c576

This video is after your last code modification. slow speed 50 high speed 100, it makes a little noise when accelerate from 50 to 100 (does not matter). But compared with right motor ( without update) last modification has more torque.

Thank you!!!

https://github.com/user-attachments/assets/be879df9-a776-4da3-afd2-30c3b7e60b58

**Comment by hoverboardhavoc** [comment]
@DDoru I also use ClassyWalk boards, and I recently updated from commit dbde6381dd2461cd99f70d692b243a3c9df0b6a0 to cdb5217e7d526a052e592c47a9b6154a9d146a76. After the update, the top speed dropped from about 450 rpm to 380 rpm. The cause appears to be the current limit being triggered at line 176 in bldc.c.

The current sensor is defined as PA6, but I believe this might be incorrect. The values fluctuate and don’t seem to reflect the actual current being drawn. I tried using the autodetect feature to check the pin, but none of the pins returned values that resemble the actual current draw.

Maybe @RoboDurden has some advice

**Comment by RoboDurden** [comment]
quickest workaround would be to not define the pin, then currentDC will stay at the initialization value of 0.42

```
	// Calculate current DC
	#ifdef CURRENT_DC
		currentDC = ABS((adc_buffer.current_dc - offsetdc) * MOTOR_AMP_CONV_DC_AMP);
	#endif
```

With autodetect, use `'m' to toggle manual mode!!!`
But if no analaog input pin shows reasonable values (reacting to chaning load on the motor), then i fear i have no idea.

**Comment by DDoru** [comment]
PA6 is just monitoring 3.3v I think, also to an test point which is not connected anywhere.

The OPAMP C32X looks that is reading the current on phase blue and green and send the information to  PA0 and PB0

![Image](https://github.com/user-attachments/assets/6107031a-72c1-4749-a4c8-762271d0c7dc)
![Image](https://github.com/user-attachments/assets/ccaa9f73-3457-4e61-8262-5c3ded26e73e)

**Comment by hoverboardhavoc** [comment]
I think that the the current `BLDC_SINE`  interpolates the rotor position based on the hall sensors. FOC uses clarke and park transformations. I think that @I-hate-2FA might have a FOC implementation? There is aslo [hoverboard-firmware-hack-FOC](https://github.com/EFeru/hoverboard-firmware-hack-FOC) which is does FOC using on Simulink generated code. It uses the ststm32 platform and stm32cube framework which are not comptible with GD32F130C8. (Maybe GD32F103 though???)

I watched all of this series a while back
https://www.youtube.com/watch?v=mbJOxqxLkLE recommend

Are these shunt resistors not connected to the phases @RoboDurden? I never saw them before they're under the heat sink

**Comment by DDoru** [comment]
> Shipping would be to Germany but i already have too much test-setups. Maybe i have the very first project for a single Gen2 board: [#92 (comment)](https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/issues/92#issuecomment-3008742993) Does Gen2.1.20 have a mpu6050 imu on board ?

I have more boards available but this board I like because never brake, they are old boards but they are ok.

I will look tomorrow but I am pretty sure that IMU will be some Chinese clone, will let you know tomorrow.

**Comment by DDoru** [comment]
IMU for this board, cannot find any information about this code, maybe the unpopulated pads on the right of IMU was for original MPU6050 :)).

![Image](https://github.com/user-attachments/assets/7f898053-e7cc-401f-b446-03c7b83343a6)

**Comment by hoverboardhavoc** [comment]
Great work @DDoru. So it looks like the two phases are measured independently?

**Comment by DDoru** [comment]
> Great work [@DDoru](https://github.com/DDoru). So it looks like the two phases are measured independently?

Yes on PA0 and PB0

**Comment by RoboDurden** [comment]
thanks for the images. I add them to https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x/tree/main/target_1%3DGD32F130/v20

**Comment by hoverboardhavoc** [comment]
Can I ask how you came by these boards @DDoru ? I also have lots of them. Some have GD32F103 (not GD32F130) processors. Do we support GD32F103 @RoboDurden? When I tried it last time there were some todos in the code

**Comment by DDoru** [comment]
> Can I ask how you came by these boards [@DDoru](https://github.com/DDoru) ? I also have lots of them. Some have GD32F103 (not GD32F130) processors. Do we support GD32F103 [@RoboDurden](https://github.com/RoboDurden)? When I tried it last time there were some todos in the code

I am repairing hoverboards, kick scooters...electric bikes..., and this old type board are form hoverboards with other problems, battery dead, motors dead... I don't throw things :))))) sometime i scavenge the MOSFET to repair other boards.....

**Comment by RoboDurden** [comment]
Thank you for your response. GD32F130  is the mcu we mostly have in these Gen2 split board hoverboards and the library files you use look just like the ones we have like: https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x-GD32/blob/main/HoverBoardGigaDevice/RTE/Device/GD32F130C8/gd32f1x0_timer.c

If you would provide a makefile, than i might be able to add your repo to my online compiler: https://pionierland.de/hoverhack/?p=test

**Comment by RoboDurden** [comment]
Ah i had a closer look at simpleFOC, and it uses an exteranal spi encoder to obtain the motor angle. So no phase currents used as we would want to do it.

**Comment by RoboDurden** [comment]
So i got a couple of 2.1.20 boards, thank you @DDoru :-)
I just uploaded my i2c scanner to a master board (that with the buzzer) but no i2c device was found on PB6/PB7 nor on on PB8/PB9
@hoverboardhavoc i am making no progress with IMU :-( I think by now i found three boards with i2c address #68 but no one could be accessed like a mpu6500.
I am close to giving up on these internal IMU.
Thanks to your code i now have basic i2c code working, so i better add a RemoteI2c with the hoverboard as the master (or slave) and a cheap imu module attached to the same i2c bus as the esp32...

**Comment by RoboDurden** [comment]
I have connected the 2.1.20 gd32f130 with such a cheap CYD:

https://github.com/user-attachments/assets/08f5c900-22df-45da-8b3a-ce247530d885

I2c bus seems to fail after some time :-(
BYD is powered from the 5V of the uart header. But powering it over USB as well does not make a difference.

Do you have pin for currentDC by now ?
Vbatt is pa4 ?

I am still testing with your mpu6050withconfig firmware 

Ideas welcome :-/

**Comment by DDoru** [comment]
> [@DDoru](https://github.com/DDoru) , the `#define MPU_6050old` works for my hoverbike, please give feedback if `#define MPU_6050 // Gemini tried for clones..` also works. I did only test this file to wake up the imu on a 2.1.4 board, which did not work. So not tested with 2.1.20

Hello, did not really understand what to do, will try Monday what you want me to test. 
I try couple of times to emulate the IMU after a sniffed the data between IMU and CPU but no luck, the hoverboard beeps 7 times...probably does not find the IMU...there is some communication between CPU and ESP32 (IMU emulator) the CPU try to read some registry.....but is beyond my knowledge. Your firmware is working ok, no problem, thank you!

**Comment by RoboDurden** [comment]
Oh i only guessed from your deleted post (which i still received an email for) that you were trying to do something with the IMU on board 2.1.20. 
If you are not, then forget about my last post. The onboard IMU is not really useful anyway.

**Comment by DDoru** [comment]
> In RemoteUart.c replace `oData.iAmpR = 0;` with
> 
> 			extern volatile uint8_t hall;
> 			oData.iAmpR = hall;

Hello, i replaced as you said, and this is data in serial monitor the same iOdom which is going up even if the motor is stopped, when you have time, thank you! 

09:33:38.934 -> millis: 102	iSpeed: 10	iSlave: 0	iOdom: 4035		iSpeed: 0.00		iAmp: 0.42		iVolt: 40.84
09:33:38.979 -> sending iSpeed: 10
09:33:39.011 -> sending iSpeed: 10
09:33:39.011 -> millis: 102	iSpeed: 10	iSlave: 0	iOdom: 4036		iSpeed: 0.00		iAmp: 0.42		iVolt: 40.84
09:33:39.045 -> sending iSpeed: 10
09:33:39.133 -> sending iSpeed: 10
09:33:39.133 -> millis: 102	iSpeed: 10	iSlave: 0	iOdom: 4037		iSpeed: 0.00		iAmp: 0.42		iVolt: 40.84  

<img width="346" height="258" alt="Image" src="https://github.com/user-attachments/assets/df928b48-5deb-4230-9297-97fbb72042db" />

**Comment by DDoru** [comment]
I try, everything, i just cannot understand. Also Gemini is too much for him  :))))
I would like to send speed forward, speed reverse, steer command and most important both motor spin same number of steps....seems you already done that with mode 3 iodom.....i watched you tube video couple of times......i just don't know how send from esp32 steering and be able to spin motors same amount of steps in order to go in straight line.
If i can get in esp32 the exact number of steps from hall sensor then i can also turn the vehicle exact 90 degree or 180 degree assuming 100% grip. If you have time and a little patience with me on my problem, because your code seems to do all i need and even more...but i just don't know how to use. Thank you!

**Comment by DDoru** [comment]
RemoteUartBus , i changed to Driving mode to 3=iOdometer, now i can send max 1000 on serial , the motors spin to 1000 ( steps i think) when i reset the motors spins back to 0. 
What to send from esp32 to spin motor 500 steps for example at 200 speed?

https://github.com/user-attachments/assets/67719271-92d2-4daa-98d9-481473f79674

**Comment by miguel-angelico** [comment]
Hi everyone,

I’m working on a university project and I’m not very experienced with this kind of firmware/tools. My hoverboard hardware appears to be Gen2.1.20. I’d like to make both motors spin at a commanded speed sent from a PC.

Setup:

    Master board MCU: GD32F103C8

    Slave board MCU: GD32F130C8

Tools: VS Code + PlatformIO (can use Keil if needed), ST-Link.

Questions:

Should I start with the REMOTE_AUTODETECT flow to identify all pins and Hall sensors/order?
My understanding is that autodetect will find LEDs, buzzer, VBATT, CURRENT_DC, SELF_HOLD, BUTTON, and HALL_A/B/C + phase order, then I should copy the printed #defines into the matching defines_2.1.20.h file and set LAYOUT in config.h. Is that the recommended first step for a Gen2.1.20 board?

Best practices for running autodetect safely:

    Power with a current-limited supply (≈1–2 A)

     Lift the wheel (no load)

In Inc/config.h:

uncomment #define REMOTE_AUTODETECT

When done, copy the printed #defines to the layout header, disable REMOTE_AUTODETECT, rebuild and test.
Is there anything else I should watch out for?


UART choice on Gen2.1.20:
For this revision, is REMOTE_USART 0 (PB6/PB7) usually the right console port, or should I prefer 1 (PA2/PA3) on the master?


Any pointers, gotchas, or a short checklist for Gen2.1.20 would be greatly appreciated. Thanks in advance!

**Comment by RoboDurden** [comment]
Isn't the 2.1.20 already read with a complete defines2_1-20.h ?
Then continue with RemoteDummy.
Always use 2A constant current power supply. Or the 1.5A charger as Power supply.

Control from PC via uart dongle and remoteUart or RemoteUartBus.

Pa2/3 is usually the master slave uart.
If you want to use this for Remote, then RemoteUartBus is your friend.

**Comment by RoboDurden** [comment]
My hoverbike is built on a 2.1.20:

https://youtu.be/ihCpCtgXIRA?si=9AQxgDJDE_qIL2L5

code and stl files online at https://github.com/RoboDurden/Hoverboard-Firmware-Hack-Gen2.x-GD32/tree/main/PlatformIO%20Examples/HoverBike%20CYD%20LVGL

Good starting point with the cheap 
Cheap Yellow Display" or CYD with touchscreen:
https://a.aliexpress.com/_EIFmgdC

And USB C port for serial connection to PC.

Pin 16 and 17 can be found on the RGB LED..

**Comment by miguel-angelico** [comment]
Today I was testing this on my master board but I’m running into some issues.

My setup:

Master board: GD32F103 (not GD32F130)

Slave board: GD32F130

I tried uploading the code to the slave (GD32F130). The upload works, but the motor doesn’t spin. On this slave board I don’t have any button. My question is: what is the purpose of SELF_HOLD and and BUTTON defined in defines_2_1_20.h?

Then I tried putting the code on the master (GD32F103). Since it has a different microcontroller, I used target 2 and created a new header file defines_2_2_20.h with the same content as defines_2_1_20.h. But when compiling in PlatformIO I get the following error:

Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/gd32/genericGD32F103C8.html
PLATFORM: GD GD32 (1.0.0+sha.b805211) > GD32F103C8 (20k RAM, 64k Flash)
HARDWARE: GD32F103C8T6 108MHz, 20KB RAM, 64KB Flash
DEBUG: Current (stlink) External (blackmagic, cmsis-dap, jlink, sipeed-rv-debugger, stlink)
PACKAGES: 
 - framework-spl-gd32 @ 2.10000.0+sha.8849cb2 
 - toolchain-gccarmnoneeabi @ 1.90201.191206 (9.2.1)
==== STARTUP FILE: startup_gd32f10x_md.S ======
SPL libraries are included: True
CMSIS libraries are included: True
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Added CMSIS-DSP library for arm_cortexM3l_math
Found 14 compatible libraries
Scanning dependencies...
No dependencies
Building in release mode
Compiling .pio\build\genericGD32F103C8\src\i2c.o
Archiving .pio\build\genericGD32F103C8\libFrameworkCMSISVariant.a
Compiling .pio\build\genericGD32F103C8\FrameworkSPL\gd32f10x_adc.o
Compiling .pio\build\genericGD32F103C8\FrameworkSPL\gd32f10x_bkp.o
Compiling .pio\build\genericGD32F103C8\FrameworkSPL\gd32f10x_can.o
Compiling .pio\build\genericGD32F103C8\FrameworkSPL\gd32f10x_crc.o
Compiling .pio\build\genericGD32F103C8\FrameworkSPL\gd32f10x_dac.o
Compiling .pio\build\genericGD32F103C8\FrameworkSPL\gd32f10x_dbg.o
Src\i2c.c: In function 'I2C_Init':
Src\i2c.c:49:3: warning: implicit declaration of function 'gpio_mode_set'; did you mean 'gpio_bit_set'? [-Wimplicit-function-declaration]
   49 |   gpio_mode_set(GPIOB, GPIO_MODE_AF, GPIO_PUPD_PULLUP, GPIO_PIN_6 | GPIO_PIN_7);
      |   ^~~~~~~~~~~~~
      |   gpio_bit_set
Src\i2c.c:49:24: error: 'GPIO_MODE_AF' undeclared (first use in this function); did you mean 'GPIO_MODE_AIN'?
   49 |   gpio_mode_set(GPIOB, GPIO_MODE_AF, GPIO_PUPD_PULLUP, GPIO_PIN_6 | GPIO_PIN_7);
      |                        ^~~~~~~~~~~~
      |                        GPIO_MODE_AIN
Src\i2c.c:50:3: warning: implicit declaration of function 'gpio_output_options_set'; did you mean 'gpio_output_port_get'? [-Wimplicit-function-declaration]
   50 |   gpio_output_options_set(GPIOB, GPIO_OTYPE_OD, GPIO_OSPEED_50MHZ, GPIO_PIN_6 | GPIO_PIN_7);
      |   ^~~~~~~~~~~~~~~~~~~~~~~
      |   gpio_output_port_get
Src\i2c.c:50:34: error: 'GPIO_OTYPE_OD' undeclared (first use in this function)
   50 |   gpio_output_options_set(GPIOB, GPIO_OTYPE_OD, GPIO_OSPEED_50MHZ, GPIO_PIN_6 | GPIO_PIN_7);
      |                                  ^~~~~~~~~~~~~
Src\i2c.c:51:3: warning: implicit declaration of function 'gpio_af_set'; did you mean 'gpio_bit_set'? [-Wimplicit-function-declaration]
   51 |   gpio_af_set(GPIOB, GPIO_AF_1, GPIO_PIN_6 | GPIO_PIN_7);  // AF1 for I2C0
      |   ^~~~~~~~~~~
      |   gpio_bit_set
Src\i2c.c:51:22: error: 'GPIO_AF_1' undeclared (first use in this function); did you mean 'GPIO_CTL1'?
   51 |   gpio_af_set(GPIOB, GPIO_AF_1, GPIO_PIN_6 | GPIO_PIN_7);  // AF1 for I2C0
      |                      ^~~~~~~~~
      |                      GPIO_CTL1
*** [.pio\build\genericGD32F103C8\src\i2c.o] Error 1
================================================================== [FAILED] Took 4.44 seconds ==================================================================


It looks like the problem is that some GPIO definitions (gpio_mode_set, GPIO_MODE_AF, gpio_af_set, etc.) don’t exist in the SPL library for GD32F103, only for the newer devices.

My questions:

What can I do to solve the error?

What’s the purpose of SELF_HOLD and BUTTON in defines_2_1_20.h? Are they required for the motor to run, or only for local start/stop logic?

Is there any recommended approach to share the same project between GD32F130 and GD32F103 targets?

Any help would be appreciated!

**Comment by DDoru** [comment]
Heloo! I am not 100% sure but my boards works like this:

On config.h  you must enable //#define DISABLE_BUTTON	for the slave board without power button and the most important part you need to check 12v 5V 3,3V on the slave board, if there is no 12v then you must add a jump wire from 12V master board to 12V slave board.

**Comment by RoboDurden** [comment]
Please copy the error log to ChatGpt or the ai you like better.

Yes, you need the 5V or 12 V pin of the master slave header to power the slave board.

Yes, a defines_2_2_20.h file should be added to the repo.

**Comment by miguel-angelico** [comment]
I’m currently testing the master branch using a GD32F103C8 board. I was able to compile successfully and run the REMOTE_AUTODETECT, which correctly detects the Hall sensors (then the LEDs change as expected using remote_dummy).

The motor spins during autodetect, but once I switch to REMOTE_DUMMY, the motor does not rotate anymore.

I’ve tried:

Testing all Hall sensor order combinations in defines.h

Printing debug variables:

bldc_enable → always active

timedOut → always 0

pos → changes correctly when I manually rotate the motor

iBldcInput, bldc_inputFilterPwm, bldc_outputFilterPwm → all update as expected

Adjusted the battery cell count as well

Despite this, the motor still doesn’t move in REMOTE_DUMMY mode.
Any idea what could be missing or what else I should check?

**Comment by RoboDurden** [comment]
Nice that you already checked what i would have suggested. 
If even bldc_outputFilterPwm goes +- 1125 then I do not really know what more to test
Try bldcBC instead of bldcSine, as this simple block communication is used in autodetect.

Also make sure that drive mode is set to 0 = simple pwm control and no fancy PID for constant speed.

**Comment by miguel-angelico** [comment]
Thanks for the suggestions!

With the help of a multimeter in AC mode, I measured the RMS voltage between two motor phases, and it always stays around 2 mV (basically zero).

Yes, bldc_outputFilterPwm goes up to about ±1125, I’m already using bldcBC instead of bldcSine, and I’ve confirmed that the driving mode is set to 0 (simple PWM control).

Could it be that the motor phase wires are not connected to the correct pins on the board? 

Still no movement from the motor. Any other ideas what could cause this?

**Comment by RoboDurden** [comment]
But autodetect makes the motor spin and the six mosfets pins in your new defines file should be the same.

You could also monitor the g y b values that are the basis for the Timer0 pwm of the six mosfet outputs.

Next I would disconnect the three motor phases , set pwm to 100% and measure the phase outputs with multimeter while slowly turning the motor to change hall pos step by step

**Comment by miguel-angelico** [comment]
I measured phase-to-phase voltage, and during Autodetect I get voltage >0V, but in RemoteDummy it’s always 0V (even when I rotate the motor manually to go through all 6 hall states).
That means the MOSFETs are not switching in RemoteDummy mode.

**Comment by miguel-angelico** [comment]
Also, could this issue be related to the SELF_HOLD or BUTTON pin configuration?
I only have one physical button on my board (the one I need to hold for the board to power on during autodetect, so I’m wondering if the other one being undefined or inactive could prevent the MOSFETs from being enabled in RemoteDummy mode (I keep checking that bldc_enable is always active anyways)

**Comment by RoboDurden** [comment]
If bldc_outputFilterPwm is fine, the motor should do something.

remoteAutodetect is ruining the same bldc.c code as RemoteDummy.

A pocket oscilloscope like https://a.aliexpress.com/_EyCqKGI or https://a.aliexpress.com/_Ez8RUrU might help.

**Comment by miguel-angelico** [comment]
Does this line 
#define TIMER_BLDC_PULLUP GPIO_PUPD_NONE
Influence something?

**Comment by RoboDurden** [comment]
i do not understand all of the gen2.x code. this is a fork of a fork.
Search for differences in defines_ad and your defines_2-20.h

I really have no more ideas.

make these variables global and monitor them with StmStudio or MCUViewer:

	int y = 0;     // yellow = phase A
	int b = 0;     // blue   = phase B
	int g = 0;     // green  = phase C

**Comment by RoboDurden** [comment]
Make a pull request with your defines_2-2-20.h then i can compile some binaries with Keil IDE for you to install with autoflash.bat (if you are on windows).

But go to bed now.

**Comment by RoboDurden** [comment]
Still in bed with my smartphone only: maybe the init function of remoteAutodetect pulls every unknown pin to output_pullup and thereby disables a protection mode that RemoteDummy does not:

```c++
void ScanInit(uint8_t iTestNew)
{
	iTest = iTestNew;
	uint8_t i;
	switch (wStage)
	{
	case AUTODETECT_Stage_Startup: return;

	....

	default:
		for (i=0;i<COUNT_PinDigital; i++)	
		{
			if (!(aoPin[i].wState & STATE_HIDE))
			{
pinModePull(aoPin[i].i,GPIO_MODE_INPUT,GPIO_PUPD_NONE);	// floating mode, no pull-up and pull-down resistors
			}
		}
	}
```

With autodetect, your motor should spin for a fraction of a second when you turn power on. 
Reduce COUNT_PinDigital down to zero to see if the motor stops spinning at startup. Like subtracting 8 each test l, then add 4 when motor stops, then subtracting 2..
Better begin with zero to see if this theory is correct.
Because it is no input_pullup but input floating..

**Comment by miguel-angelico** [comment]
It works now! 🎉

The issue was exactly this line:

#define TIMER_BLDC_PULLUP GPIO_MODE_AF_PP


I had it set as GPIO_PUPD_NONE, so the Timer0 PWM signals weren’t actually reaching the MOSFET driver.
After changing it to GPIO_MODE_AF_PP, the PWM outputs are active and the motor spins perfectly in RemoteDummy mode.

Thanks a lot for your help and patience. I’m really happy it’s finally working!

**Comment by RoboDurden** [comment]
Please upload photos of the f103 board, so I can add it as 2.2.20 to the interactive board finder.
